### PR TITLE
ci: use sidecar versions from the build.env 

### PR DESCRIPTION
**Pull request by humblec**
The sidecar images in minikube deployment will be fetched from build.env and used/validated accordingly.

Testing ( with all versions set):
```
validating sidecar controller configuration
```
Testing without a sidecar version set:
```
validating sidecar controller configuration
CSI_RESIZER_VERSION version is empty, make sure build.env has set this sidecar version

```

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Review by Madhu-1 — COMMENTED** [review]
This is supposed to be overridden from the build.env, if we start updating sidecar is both places it's a double work. We should stick to updating it in only one place.

**Comment by humblec** [comment]
> This is supposed to be overridden from the build.env, if we start updating sidecar is both places it's a double work. We should stick to updating it in only one place.

Yeah, however its is possible to use this `scripts/minikube.sh` locally  , where the version update make sense. 

As we keep the versions in the script its indeed better to keep the latest versions in place. Also sidecar updates are not going to happen very frequently.. Mostly once with a kube release.  iow, there is no point to keep very old versions in the script either..

**Review by Madhu-1 — COMMENTED** [review]
Do source build.env in minikube.sh and use it? and this avoids duplicate works if we maintain image versios in multiple places

**Comment by humblec** [comment]
See sourcing can be used, but my point is simple, if we keep versions, keep latest versions in place, and if we dont want to maintain the versions separately in this file,  there is no point to keep the very old versions out there. Also sourcing make the dependency to build.env , if thats the case we dont even need mention of these versions here in this file.

**Comment by Madhu-1** [comment]
> See sourcing can be used, but my point is simple, if we keep versions, keep latest versions in place, and if we dont want to maintain the versions separately in this file, there is no point to keep the very old versions out there. Also sourcing make the dependency to build.env , if thats the case we dont even need mention of these versions here in this file.

I don't like the idea of updating it in multiple places which is completely useless. if you think it can be removed please remove and add a validation to check the image versions are set.

**Comment by humblec** [comment]
So, locally it works, however in CI, the `build.env` path is different... any thoughts  here @ceph/ceph-csi-contributors

**Comment by Madhu-1** [comment]
> So, locally it works, however in CI, the `build.env` path is different... any thoughts here @ceph/ceph-csi-contributors

source build.env like below?

```bash
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
# shellcheck source=build.env
source "${SCRIPT_DIR}/../build.env"
```

**Comment by humblec** [comment]
> > So, locally it works, however in CI, the `build.env` path is different... any thoughts here @ceph/ceph-csi-contributors
> 
> source build.env like below?
> 
> ```
> SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
> # shellcheck source=build.env
> source "${SCRIPT_DIR}/../build.env"
> ```

I knew this existed some where but was not able to fetch it .. Thanks.. should be good now. :+1:

**Review by nixpanic — CHANGES_REQUESTED** [review]
The PR description and commit message suggest that this is for deploying (by Helm or `deploy/` yaml files). However this is for CI only?

Maybe it would be good to validate that the deployment files actually have the same versions of the sidecars as listed in `build.env`. Currently the versions in `build.env` may be different in `charts/` and `deploy/`.

**Comment by humblec** [comment]
> The PR description and commit message suggest that this is for deploying (by Helm or `deploy/` yaml files). However this is for CI only?
> 
> Maybe it would be good to validate that the deployment files actually have the same versions of the sidecars as listed in `build.env`. Currently the versions in `build.env` may be different in `charts/` and `deploy/`.

I will keep this PR for CI.. Updated commit and PR description.

**Comment by humblec** [comment]
@nixpanic kept this PR for CI only and updated. ptal.. thanks.

File: scripts/minikube.sh [hunk 1]
```diff
@@ -162,11 +162,21 @@ else
...
-CSI_SNAPSHOTTER_VERSION=${CSI_SNAPSHOTTER_VERSION:-"v4.1.1"}
-CSI_PROVISIONER_VERSION=${CSI_PROVISIONER_VERSION:-"v2.2.2"}
-CSI_RESIZER_VERSION=${CSI_RESIZER_VERSION:-"v1.2.0"}
-CSI_NODE_DRIVER_REGISTRAR_VERSION=${CSI_NODE_DRIVER_REGISTRAR_VERSION:-"v2.2.0"}
+
+# as we are no longer keeping the sidecar versions here, try to fetch it from build.env file
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+source "${SCRIPT_DIR}/../build.env"
+
+sidecars=(CSI_ATTACHER_VERSION CSI_SNAPSHOTTER_VERSION CSI_PROVISIONER_VERSION CSI_RESIZER_VERSION CSI_NODE_DRIVER_REGISTRAR_VERSION)
```
Comment (Madhu-1) [inline scripts/minikube.sh]: I don't think we need to check the sidecars version by default, this script is used for many tasks. creating minikube cluster. deploying Rook. copying sidecar images. creating pools etc. this check should only be called if someone is using this to copy the sidecar images.

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

File: scripts/minikube.sh [hunk 2]
```diff
@@ -162,11 +162,21 @@ else
...
+for controller in "${sidecars[@]}"; do
+	if [[ "${!controller}" ]]; then
```
Comment (nixpanic) [inline scripts/minikube.sh]: For checking empty strings, use `-z`.
No need for listing the versions, that is just clutter IMHO.
This check is not always needed, only for deploying?

File: scripts/minikube.sh [hunk 3]
```diff
@@ -162,11 +162,21 @@ else
...
+		echo "${controller}" : "${!controller}"
+	else
+		echo "${controller}" version is empty, make sure build.env has set this sidecar version
+               exit 1
```
Comment (nixpanic) [inline scripts/minikube.sh]: indention needs to be fixed

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: scripts/minikube.sh [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash -e
-
+# shellcheck disable=SC1091
```
Comment (nixpanic) [inline scripts/minikube.sh]: please only do this where needed, not for the whole script

**Review by nixpanic — DISMISSED** [review]

File: scripts/minikube.sh (hunk 1 above)
Comment (humblec) [reply scripts/minikube.sh]: Done.

File: scripts/minikube.sh (hunk 2 above)
Comment (humblec) [reply scripts/minikube.sh]: sure, used `-z`

File: scripts/minikube.sh (hunk 4 above)
Comment (humblec) [reply scripts/minikube.sh]: Done.

**Comment by humblec** [comment]
@nixpanic @Madhu-1 all comments are addressed. ptal .. thanks.

File: scripts/minikube.sh [hunk 5]
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
...
 }
 
+# configure csi sidecar version
+# shellcheck disable=SC1091
+function validate_sidecar() {
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+source "${SCRIPT_DIR}/../build.env"
+
+sidecars=(CSI_ATTACHER_VERSION CSI_SNAPSHOTTER_VERSION CSI_PROVISIONER_VERSION CSI_RESIZER_VERSION CSI_NODE_DRIVER_REGISTRAR_VERSION)
+for controller in "${sidecars[@]}"; do
```
Comment (Madhu-1) [inline scripts/minikube.sh]: use sidecar instead of controller for variable name?

File: scripts/minikube.sh [hunk 6]
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
     fi
 }
 
+# configure csi sidecar version
+# shellcheck disable=SC1091
+function validate_sidecar() {
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
```
Comment (Madhu-1) [inline scripts/minikube.sh]: function indentation is wrong compared to other functions in the file

File: scripts/minikube.sh [hunk 7]
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
     fi
 }
 
+# configure csi sidecar version
+# shellcheck disable=SC1091
```
Comment (Madhu-1) [inline scripts/minikube.sh]: why disable check for the whole function?

**Comment by humblec** [comment]
@nixpanic @Madhu-1 comments are addressed. ptal. thanks.

**Comment by humblec** [comment]
@Madhu-1 @nixpanic ptal. thanks.

**Comment by humblec** [comment]
@Madhu-1 @nixpanic can you please revisit this PR as you have changes requested here?

File: scripts/minikube.sh [hunk 8]
```diff
@@ -290,6 +298,8 @@ cephcsi)
     copy_image_to_cluster "${CEPHCSI_IMAGE_REPO}"/cephcsi:"${CSI_IMAGE_VERSION}" "${CEPHCSI_IMAGE_REPO}"/cephcsi:"${CSI_IMAGE_VERSION}"
     ;;
 k8s-sidecar)
+    echo "validating sidecar controller configuration"
```
Comment (Madhu-1) [inline scripts/minikube.sh]: `echo "validating sidecar controller configuration"` to  `echo "validating sidecar images version"`

File: scripts/minikube.sh [hunk 9]
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
     fi
 }
 
+# configure csi sidecar version
+function validate_sidecar() {
```
Comment (Madhu-1) [inline scripts/minikube.sh:128]: comment and function name is not matching rename this to configure_sidecar or update the comment?

File: scripts/minikube.sh [hunk 10]
```diff
@@ -124,6 +124,21 @@ function validate_container_cmd() {
...
+    SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+# shellcheck disable=SC1091
```
Comment (Madhu-1) [inline scripts/minikube.sh:130]: the comment is not indented is this expected?

**Review by Madhu-1 — COMMENTED** [review]
LGTM. left few suggestions

**Comment by humblec** [comment]
@Rakshith-R @yati1998 ptal.. thanks

**Review by yati1998 — APPROVED** [review]
LGTM

**Review by Rakshith-R — APPROVED** [review]

**Comment by humblec** [comment]
@Madhu-1 ptal.. thanks.

**Review by Madhu-1 — APPROVED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by Rakshith-R** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio refresh

**Comment by humblec** [comment]
@Mergifyio rebase

**Pull request merged**